### PR TITLE
Fix 97272

### LINKED ISSUE #97272: Assertion failed '!"Unhandled must expand intrinsic, throwing PlatformNotSupportedException"' in Checked mode
Found in https://github.com/dotnet/runtime/pull/97087#issuecomment-1901068671, only happens in Checked build.

> Generating native code
ILC: D:\a_work\1\s\src\coreclr\jit\importercalls.cpp:4106
ILC: Assertion failed '!"Unhandled must expand intrinsic, throwing PlatformNotSupportedException"' in 'System.Type:get_IsPrimitive():ubyte:this' during 'Morph - Inlining' (IL size 7; hash 0x30bca526; FullOpts)

Likely caused by https://github.com/dotnet/runtime/pull/96226

Happens in:
- DisabledRuntimeMarshalling_Disabled_NativeAssemblyEnabled.csproj
- AmbiguousImplementationException.ilproj
- Pause_ro.csproj
- Regression2_Regressions.csproj

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes https://github.com/dotnet/runtime/issues/97272

### CODE REVIEW

**Comment by EgorBo** [comment]
@MihuBot

**Comment by Sergio0694** [comment]
MrBeanWaitingGIF

**Comment by Sergio0694** [comment]
Me checking this PR before going to bed:

SchooldaysGIF

**Comment by Sergio0694** [comment]
WaitingSpongebobGIF

**Comment by Sergio0694** [comment]
CatLayingGIF

**Comment by Sergio0694** [comment]
Egor plz

BabyCryingBabyGIF

**Comment by EgorBo** [comment]
> Egor plz
> 
> BabyCryingBabyGIF [ BabyCryingBabyGIF ](https://private-user-images.githubusercontent.com/10199417/329053109-2e07e85f-99dd-4122-9e2f-c3a674f4538c.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUyMDM0OTMsIm5iZiI6MTcxNTIwMzE5MywicGF0aCI6Ii8xMDE5OTQxNy8zMjkwNTMxMDktMmUwN2U4NWYtOTlkZC00MTIyLTllMmYtYzNhNjc0ZjQ1MzhjLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA4VDIxMTk1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJiMjk2NzNmZGZkOWUyMWMwNjY0YmI0OTYyYmU0OTdiOWQ3NDBmMjU0MTRlNTdjNGQ3MGYzMzdlNjQ1NGI3NzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.SPq9VCR0g-1MTZjjYQ4O7votjJVUwZC6mqylulva-QM) [ ](https://private-user-images.githubusercontent.com/10199417/329053109-2e07e85f-99dd-4122-9e2f-c3a674f4538c.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUyMDM0OTMsIm5iZiI6MTcxNTIwMzE5MywicGF0aCI6Ii8xMDE5OTQxNy8zMjkwNTMxMDktMmUwN2U4NWYtOTlkZC00MTIyLTllMmYtYzNhNjc0ZjQ1MzhjLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA4VDIxMTk1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJiMjk2NzNmZGZkOWUyMWMwNjY0YmI0OTYyYmU0OTdiOWQ3NDBmMjU0MTRlNTdjNGQ3MGYzMzdlNjQ1NGI3NzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.SPq9VCR0g-1MTZjjYQ4O7votjJVUwZC6mqylulva-QM)

haha, sorry Sergio it fell out from my radar due to high priority things. I think for this I only need to inspect the SPMI diffs for legality and ask someone for review

**Comment by EgorBo** [comment]
@tannergooding or anyone from @dotnet/jit-contrib could you take a look? 
tl;dr: we currently detect must-expand intrinsics as "does this call invoke current root method" while it should be - "does this call invoke itself" regardless of who is root - this led to problems with [Intrinsic] attribute on top of virtual methods in `Type`. The interesting side-effect from this is that we now can inline must-expan SIMD intrsinsics which we give up due to non-constant args for IMM etc. It allows jit to fold the fallback if late phases detect a constant. Example:

```cs
Vector128<int> Foo(Vector128<int> v)
{
    byte mask = 0b11110000;
    return Sse2.Shuffle(v, mask);
}
```

Main:
```asm
; Method MyBench:Foo
       mov      rcx, rdx
       mov      rdx, r8
       mov      r8d, 240
       tail.jmp [Sse2:Shuffle(Vector128`1[int],ubyte):Vector128`1[int]]
; Total bytes of code: 18
```
PR:
```asm
; Method MyBench:Foo
       vpshufd  xmm0, xmmword ptr [r8], -16
       vmovups  xmmword ptr [rdx], xmm0
       mov      rax, rdx
       ret      
; Total bytes of code: 14
```

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=682093&view=ms.vss-build-web.run-extensions-tab)

There a few regressions which are just PMI-artifacts, e.g. we inline the fallback [here](https://github.com/dotnet/runtime/blob/17e6a5f8a4f18d64d07c90c76e7e81736fff3f92/src/libraries/System.IO.Hashing/src/System/IO/Hashing/VectorHelper.cs#L105). but it shouldn't matter since the method is supposed to be inlined (AggressiveInlining).

+ a couple of r2r regressions - I investigated those and looks like they're "cancel inlining because `BlockNonDeterministicIntrinsics` said so"

**Comment by tannergooding** [comment]
There's two libraries-pmi regressions that look interesting. I think in production they aren't actually hit for the cases highlighted because they are small methods that get inlined and the constant is propagated.

But, it showcases an issue where we aren't preserving it as a call for the non-constant case, so we end up expanding the jump table inline instead, this can lead to nasty code explosion in some cases and that is potentially concerning (particularly for T0 like code).

File: src/coreclr/jit/hwintrinsic.cpp [hunk 1]
```diff
@@ -1076,9 +1076,8 @@ bool Compiler::CheckHWIntrinsicImmRange(NamedIntrinsic intrinsic,
 
         if (immOutOfRange)
         {
-            assert(!mustExpand);
-            // The imm-HWintrinsics that do not accept all imm8 values may throw
-            // ArgumentOutOfRangeException when the imm argument is not in the valid range
+            // We'll use a fallback throwing ArgumentOutOfRangeException
+            *useFallback = true;
```
Comment (tannergooding) [inline src/coreclr/jit/hwintrinsic.cpp:1080]: Are we losing or missing context here that the call will definitively throw? Maybe that can help with control flow and tagging things as hot/cold?

File: src/coreclr/jit/importercalls.cpp [hunk 2]
```diff
@@ -3068,9 +3068,10 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 }
             }
 
-            GenTree* hwintrinsic = impHWIntrinsic(ni, clsHnd, method, sig, mustExpand);
+            bool     importAsCall = false;
+            GenTree* hwintrinsic  = impHWIntrinsic(ni, clsHnd, method, sig, mustExpand, &importAsCall);
 
-            if (mustExpand && (hwintrinsic == nullptr))
+            if (mustExpand && (hwintrinsic == nullptr) && !importAsCall)
```
Comment (tannergooding) [inline src/coreclr/jit/importercalls.cpp:3074]: When would we be in a scenario where we're `mustExpand` and `hwintrinsic == nullptr` and `importAsCall` is true?

This path is basically supposed to be ensuring we never end up with infinite recursion and therefore stack overflow and it seems like the new condition would end up allowing that to exist.

File: src/coreclr/jit/importercalls.cpp [hunk 3]
```diff
@@ -10062,7 +10063,7 @@ NamedIntrinsic Compiler::lookupNamedIntrinsic(CORINFO_METHOD_HANDLE method)
                                 assert(strcmp(className, "Vector`1") == 0);
                                 result = NI_Vector_GetCount;
                             }
-                            else if (gtIsRecursiveCall(method))
```
Comment (tannergooding) [inline src/coreclr/jit/importercalls.cpp:10065]: Is this a general issue around all `gtIsRecursiveCall` that should maybe be fixed? The intent for that API was basically "am I calling myself" and calling some base or derived virtual method target shouldn't count as recursion.

File: src/coreclr/jit/hwintrinsic.cpp [hunk 4]
```diff
@@ -1310,6 +1318,13 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
         if (!CheckHWIntrinsicImmRange(intrinsic, simdBaseJitType, immOp1, mustExpand, immLowerBound, immUpperBound,
                                       hasFullRangeImm, &useFallback))
         {
+            // If it needs a fallback - we have two options: Importer-level via impNonConstFallback and lower-level
+            if (useFallback && !HWIntrinsicInfo::NoJmpTableImm(intrinsic) &&
+                !HWIntrinsicInfo::MaybeNoJmpTableImm(intrinsic))
+            {
+                *pImportAsCall = true;
```
Comment (tannergooding) [inline src/coreclr/jit/hwintrinsic.cpp:1325]: I'm not following where the logic that allows the call to be converted back into an intrinsic is?

That is, I'm not following the logic that allows this:
> The interesting side-effect from this is that we now can inline must-expand SIMD intrsinsics which we give up due to non-constant args for IMM etc. It allows jit to fold the fallback if late phases detect a constant

I would've expected either we import the `GT_HWINTRINSIC` node and rewrite it back to a `GT_CALL` in a late phase (such as rationalization). Or we instead import as `GT_CALL` and then have a later phase (morph and/or rationalization) attempt to resolve the tracked method handled into an intrinsic again (repeating the checks that the right input is a constant).

**Comment by Sergio0694** [comment]
IMissYouGIF

Egor pls

**Comment by EgorBo** [comment]
@tannergooding PTAL again, your work to rewrite HWINTRINSIC as calls helped to remove the hacks I had in the previous version.

File: src/coreclr/jit/importercalls.cpp (hunk 3 above)
Comment (EgorBo) [reply src/coreclr/jit/importercalls.cpp:10138]: the current impl of `gtIsRecursiveCall` is "am I calling whatever is root method now" - it's fine for most cases, but in case of must-expand intrinsics, we only care about "method intentionally invoking itself"

**Review by tannergooding — APPROVED** [review]
This looks a lot better now, thanks!

**Comment by Sergio0694** [comment]
finished-elijah-wood
